docs/implementing-new-trigger.md

```diff
@@ -0,0 +1,71 @@
+
+For e.g below is the definition of [Kafka Trigger](https://github.com/kubeless/kubeless/blob/master/pkg/apis/kubeless/v1beta1/kafka_trigger.go)
+
+```go
+type KafkaTrigger struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              KafkaTriggerSpec `json:"spec"`
+}
+```
+
+You need to model the event source attributes in to Spec attribute of new trigger. Dependeing on the nature of event source you may want to associate single function or multiple functions with the event source. Use appropriate mechanism to represent the association. For e.g Kafka trigger uses Kubernetes label selector to associate any function with matching label with the event source.
+
+## Code Generation
+
+Once you have definged the new trigger, please ensure its placed in `pkg/apis/kubeless/v1beta1/` path, and update `register.go` go to include new trigger type.
+
+Now you can auto-generate the clientset, lister and informers for the new API resource object as well but running `make update` or `./hack/update-codegen.sh`. Auto-generated clientset, lister and informers comes handy in writing the controller in next step.
+
+## CRD controller
+
+Here is the most important step, i.e. writing controller itself. As far as the skelton of controller goes, it would be pretty similar to existing controllers like Kafka trigger controller, nats trigger controller or http trigger controller. Functionally controller does two important things
+
+- watch Kuberentes API server for CRUD operations on the new trigger object and take appropriate actions.
+- when an event occurs in the event source trigger the associated functions.
+
+Please read the code and logic for existing [controllers](https://github.com/kubeless/kubeless/tree/master/pkg/controller) as a referance.
+
+## Building controller binary and docker image
+
+Ensure you controller is independent binary that can be built from the Makefile. Please follow one of the existing controller [cmd](https://github.com/kubeless/kubeless/tree/master/cmd) as referance. Also ensure there is corresponding `Dockerfile` to build the controller image. Plese see the dockerfile for other trigger controller as a [referance](https://github.com/kubeless/kubeless/tree/master/docker).
+
+Add apparopriate Makefile targets so that controller binary and docker image can be built.
+
+## Manifest
+
+Create a jsonnet file for the new trigger and ensure that generated yaml file has CRD definition, deployment for the trigger controller and necessary RBAC rules. Again most of the stuff is common with Kafka or HTTP triggers, so take existing jsonnet manifests as a referance.
+
+## CI
+Please add unit tests and integration tests to tests the new functionality related to adding a trigger in to Kubeless. Also make necessary changes to .travis.yaml so that new trigger controller image can be built, deployed and integration tests can be run against the new trigger functionality.
```

Comment:
It'd be great to have a checklist of the minimum things that the integration tests should cover.